### ISSUE TITLE
Constrained 2D Delaunay triangulation : duplicate point leads to degenerated triangles

### ISSUE THREAD

**Issue by yobeonline** [CGAL/cgal#3261]
## Issue Details
I am inserting three sets of constraints that each form a triangle (see the picture below).
![cgal](https://user-images.githubusercontent.com/6747553/43389632-773206e0-93ec-11e8-8c8a-e2acaf454229.png)

The problem is around p4 (which is not included in the green triangle) : the produced triangulation creates many instances of this point, leading to faces that are degenerated.

## Source Code

Here is the source code that reproduces the problem:

```c++
#include <iostream>

#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Constrained_Delaunay_triangulation_2.h>
#include <CGAL/Constrained_triangulation_plus_2.h>
#include <CGAL/Triangulation_face_base_2.h>
#include <CGAL/Triangulation_vertex_base_2.h>
#include <CGAL/Delaunay_triangulation_2.h>

using K = CGAL::Exact_predicates_inexact_constructions_kernel;

using Vb2 = CGAL::Triangulation_vertex_base_2<K>;
using Fb2 = CGAL::Constrained_triangulation_face_base_2<K>;

using TDS2 = CGAL::Triangulation_data_structure_2<Vb2,Fb2>;
using Itag2 = CGAL::Exact_predicates_tag;
using CDT2_Base = CGAL::Constrained_Delaunay_triangulation_2<K,TDS2,Itag2>;

using Triangulation = CGAL::Constrained_triangulation_plus_2<CDT2_Base>;
using point = Triangulation::Point;

void main(void)
{
  Triangulation t;

  point const p1(381886.770019531, 3949942.71002197);
  point const p2(381904.979980469, 3949910.38000488);
  point const p3(381923.260009766, 3949963.55999756);
  point const p4(381889.210021973, 3949938.38000488);
  point const p5(381871.590026855, 3949928.2800293);

  { // green : doesn't contain p4
    t.insert_constraint(p1,p2);
    t.insert_constraint(p2,p3);
    t.insert_constraint(p3,p1);
  }

  { // blue
    t.insert_constraint(p5,p2);
    t.insert_constraint(p2,p4);
    t.insert_constraint(p4,p5);
  }

  { // yellow
    t.insert_constraint(p5,p4);
    t.insert_constraint(p4,p1);
    t.insert_constraint(p1,p5);
  }

  std::ofstream f("triangulation.txt");
  std::cout << "number of faces: " << t.number_of_faces() << '\n'; // prints 12
  std::cout << "number fo vertices: " << t.number_of_vertices() << '\n'; // prints 9

  std::cout.precision(12);
  std::cout << "triangulation:\n" << t; // contains four instances of 381889.209183 3949938.37952
                                             // plus one of 381889.210022 3949938.38
  return;
}

```
The output of this function is:

> number of faces: 12
number fo vertices: 9
triangulation:
10 16 2
381886.77002 3949942.71002
381904.97998 3949910.38
381923.26001 3949963.56
381871.590027 3949928.28003
381889.210022 3949938.38
381889.209183 3949938.37952
381889.209183 3949938.37952
381889.209183 3949938.37952
381889.209183 3949938.37952
>
>2 0 3 
5 2 3 
8 5 7 
0 1 3 
4 1 0 
2 4 0 
1 5 3 
8 4 2 
7 6 8 
9 4 6 
1 7 5 
1 9 7 
5 8 2 
6 4 8 
7 9 6 
1 4 9 
>
>3 1 5 
0 6 12 
10 8 12 
6 0 4 
3 5 15 
4 0 7 
1 3 10 
5 12 13 
13 2 14 
13 14 15 
2 6 11 
14 10 15 
7 1 2 
7 8 9 
9 8 11 
9 11 4 
N C N
C N C
C C C
C N N
N N C
N N C
N C C
C C N
C C C
C C C
C C N
C N C
C C C
N C C
C C C
C C C

I expect the result to be the one I am getting if I manually insert p4 in the green triangle: (probably not possible due to rounding errors)
```c++
{ // green : manually inserting p4
    t.insert_constraint(p1,p4);
    t.insert_constraint(p4,p2);
    t.insert_constraint(p2,p3);
    t.insert_constraint(p3,p1);
}

```

Which prints:

> number of faces: 4
number fo vertices: 5
triangulation:
6 8 2
381886.77002 3949942.71002
381889.210022 3949938.38
381904.97998 3949910.38
381923.26001 3949963.56
381871.590027 3949928.28003
>
>0 4 3 
2 1 5 
1 2 4 
5 0 3 
4 2 3 
1 4 0 
2 5 3 
1 0 5 
>
>4 3 5 
7 6 2 
4 5 1 
0 6 7 
6 0 2 
0 7 2 
3 4 1 
3 1 5 
C N N
C C C
N C C
N C N
C C N
N N C
C C C
N C N

## Environment

* Windows 64 bits
* Compiler: vs 2013 32 and 64
* Release or debug mode: both
* Specific flags used (if any): none
* CGAL version: 4.11.2
* Boost version: 1.65.1
* Other libraries versions if used (Eigen, TBB, etc.): none

**Comment by lrineau** [comment]
The issue is because the segment [p1, p4] intersects [p2, p3]. In CGAL-4.11.2, the behavior of unexact constructions lead to a cascade of intersections, hence the 9 vertices at the end. I just released a new version of CGAL, version 4.11.3, that tweaks the constructions to avoid creating points that are really close to each other (< 4 [ULP]). With that version of CGAL, your input data produces a triangulation with six vertices. The sixth vertex is the intersection of [p1, p4] with [p2, p3], that is close to p4, but at a distance a lot more than 4 ULP.

[ULP]: https://en.wikipedia.org/wiki/Unit_in_the_last_place

**Comment by yobeonline** [comment]
I could test the release 4.11.3 and it fixes it indeed. Sorry for the noise of my defect submissions.

I have one or two more on their way though: one of the same nature that is not fixed by 4.11.3 (which I can't reproduce in isolation at the moment) and another, more serious (infinite loop?) that have been introduced by the 4.11.3 release. The latter is possibly linked to the first one since it occures in the very code that was attempting to reproduce it.
